### ISSUE TITLE
usePresence throws error if Ably is not configured

### ISSUE THREAD

**Issue by danielrhodes** [ably-labs/react-hooks#47]
If `usePresence` or `useChannel` is used without an internet connection (and Ably has not been activated yet), I get this error:

```Error: Ably not configured - please call configureAbly({ key: "your-api-key", clientId: "someid" });``

While I'm sure this might technically be true, it is coming from `assertConfiguration`. My expectation is this will fail silently in this scenario.

In general, it is bad form to throw an error inside a hook because it can result in an inconsistent state in the app. Instead it would be preferable to return functions which throw errors - that way the errors can be caught appropriately and dealt with.

**Comment by danielrhodes** [comment]
After thinking about this a little more, I think the appropriate way to deal with this is to have a context provider that holds the client. I'd suggest taking a look at how Apollo has done this (which also needs to maintain a WS connection).

**Comment by owenpearson** [comment]
Hey @danielrhodes, thanks for the feedback!

I agree that using a context provider is a better approach for hooks which require a stateful connection. Unfortunately we wouldn't be able to do this without a breaking change release, I've created https://github.com/ably-labs/react-hooks/issues/49 to track that separately.

As for throwing an error from the hook - I can see that it causes problems, although on the other hand, this error will only be thrown if the hook is missing a critical dependency - in your case it sounds it may make more sense to avoid rendering a component with `usePresence` or `useChannel` hooks until the client has been created, although I appreciate that may not always be easily possible. Out of interest do you have any examples of hook libraries which handle errors by returning functions?

**Comment by danielrhodes** [comment]
> in your case it sounds it may make more sense to avoid rendering a component with usePresence or useChannel hooks until the client has been created

Yep that is the problem and breaks assumptions about how hooks work, which is that they don't conditionally render.

> Out of interest do you have any examples of hook libraries which handle errors by returning functions?

Not sure what you mean by returning a function. But in terms of errors, you would return an error object that can be consumed by the child.

I'd recommend looking at https://github.com/apollographql/apollo-client as a good example.

**Comment by danielrhodes** [comment]
@owenpearson This hasn't been tested much, but I rewrote the hooks using. the provider. https://gist.github.com/danielrhodes/7271ad41eb3d189e62f58309a6b5961c

```ts
import {Realtime} from 'ably';

const RootComponent = React.memo(() => {
  const client = useMemo(() => {
    if (someVarLikeAToken) {
      console.log('Creating a new one!');
      const realtime = new Realtime({
         ...
      });
      return realtime;
    } else {
      return null;
    }
  }, [someVarLikeAToken]);

  return (
    <AblyProvider client={client}>
      <MoreAppStuff />
    </AblyProvider>
  );
});
```

**Comment by WillSmithTE** [comment]
This is cool, just chiming in to say I'd be really happy to switch to a context provider

**Comment by owenpearson** [comment]
Hey @danielrhodes, thanks for sharing that gist 🙂 I've opened a PR to replace `configureAbly` with a context provider, the implementation is similarly based off of apollo client but we needed some changes because while apollo enforces a single context per page, there are legitimate reasons to use multiple ably clients in a single react app. Anyway, this should be available in our next version react-hooks, thanks again for the suggestion!

**Comment by danielrhodes** [comment]
@owenpearson Thanks for jumping on this so quickly. I am going to move some comments I have to the PR discussion.

**Comment by owenpearson** [comment]
Closing this issue for now as we've implemented `AblyProvider` in #57
